taichi/backends/opengl/opengl_program.cpp

```diff
@@ -32,6 +32,17 @@ void OpenglProgramImpl::materialize_runtime(MemoryPool *memory_pool,
   TI_NOT_IMPLEMENTED;
 #endif
 }
+DeviceAllocation OpenglProgramImpl::allocate_memory_ndarray(
+    std::size_t alloc_size,
+    uint64 *result_buffer) {
+  return opengl_runtime_->device->allocate_memory(
+      {alloc_size, /*host_write=*/true, /*host_read=*/true,
```

Comment:
These two flags should be used carefully since it not only declares a capability, it also hints the device backend to place this memory into different memory heaps (with potentially degraded on-gpu performance or host-device sync overhead)

Reply:
@bobcao3 These were mostly from https://github.com/taichi-dev/taichi/blob/v0.8.4/taichi/backends/opengl/opengl_api.cpp#L374. But I did try setting `host_read` to false and it also passed tests. (setting `host_write` to false will break tests) But I'm not 100% if I understood `host_read` correctly. Do you have any insights here?

Reply:
i'm also curious here since in our [CUDA device allocate_memory](https://github.com/taichi-dev/taichi/blob/46b27c7c2f0970ec822c6a197f8c50f9bab6318b/taichi/backends/cuda/cuda_device.cpp#L16), these two flags lead to the usage of `malloc_managed` which hints unified memory, which has been dropped out?

Reply:
>  setting host_read to false and it also passed tests.

https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glMapBufferRange.xhtml My feeling is that WRITE implies the memory is readable.

Anyway, we can do fancier analysis later. For now it's safe to turn on both read and write.

> these two flags lead to the usage of malloc_managed which hints unified memory, which has been dropped out?

I don't think we have unified the memory allocation mechanism yet. For now the one goes through the preallocated memory is in https://github.com/taichi-dev/taichi/blob/7027cc300cf459f57dcc882dd5ddcae14ae7281e/taichi/backends/cuda/cuda_device.cpp#L38-L44